framework/src/timesteppers/IterationAdaptivePostprocessorDT.C

```diff
@@ -0,0 +1,526 @@
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+// MOOSE includes
+#include "IterationAdaptivePostprocessorDT.h"
+#include "Function.h"
+#include "Piecewise.h"
+#include "Transient.h"
+#include "NonlinearSystem.h"
+
+template<>
+InputParameters validParams<IterationAdaptivePostprocessorDT>()
+{
+  InputParameters params = validParams<TimeStepper>();
+  params.addParam<int>("optimal_iterations", "The target number of nonlinear iterations for adaptive timestepping");
+  params.addParam<int>("iteration_window",  "The size of the nonlinear iteration window for adaptive timestepping (default = 0.2*optimal_iterations)");
+  params.addParam<unsigned> ("linear_iteration_ratio", "The ratio of linear to nonlinear iterations to determine target linear iterations and window for adaptive timestepping (default = 25)");
+  params.addParam<FunctionName> ("timestep_limiting_function", "A function used to control the timestep by limiting the change in the function over a timestep");
+  params.addParam<Real> ("max_function_change", "The absolute value of the maximum change in timestep_limiting_function over a timestep");
+  params.addParam<bool> ("force_step_every_function_point", false,"Forces the timestepper to take a step that is consistent with points defined in the function.");
+  params.addRequiredParam<Real>("dt", "The default timestep size between solves");
+  params.addParam<std::vector<Real> >("time_t", "The values of t");
+  params.addParam<std::vector<Real> >("time_dt", "The values of dt");
+  params.addParam<Real>("growth_factor",   2, "Factor to apply to timestep if easy convergence (if 'optimal_iterations' is specified) or if recovering from failed solve");
+  params.addParam<Real>("cutback_factor",  0.5, "Factor to apply to timestep if difficult convergence (if 'optimal_iterations' is specified) or if solution failed.");
+  params.addRequiredParam<PostprocessorName>("postprocessor", "The name of the postprocessor that computes the dt");
+
+  return params;
+}
+
+IterationAdaptivePostprocessorDT::IterationAdaptivePostprocessorDT(const InputParameters & parameters) :
+  TimeStepper(parameters),
+  PostprocessorInterface(parameters),
+  _dt_old(declareRestartableData<Real>("dt_old",0.0)),
+  _input_dt(getParam<Real>("dt")),
+  _tfunc_last_step(declareRestartableData<bool>("tfunc_last_step",false)),
+  _sync_last_step(declareRestartableData<bool>("sync_last_step",false)),
```

Comment:
Any restartable data type needs to be declared as a reference. I see several here that are not references in your header file. These won't work correctly with restart which makes me think you don't need them marked as restartable.